### PR TITLE
Implement an interface to import PDF metadata from multiple sources (XMP, Grobid, ...)

### PR DESCRIPTION
**Pull request by btut**
We want to be able to import PDFs into JabRef and infer all Bib-data from the file itself.
This is done by using multiple importers. If they disagree about the metadata, we need a way to merge the conflicting data. As of #7947, this is done by prioritization of importers.
For pro-users, we want to have the option to do the merge manually using an n-way merge dialog. It can be triggered by clicking a button next to a linked (offline, pdf) file.
Screenshot from 2021-08-16 12-34-56

The dialog looks like a table. Each source will be represented by a column, each field by a row. There will be an additional, editable, column that represents the final entry.

Peek 2021-08-17 14-00

Users can:
- Enter the information manually in the final entry column
- Select one source to copy all it's fields to the final entry column
- Select a field to copy it's content to the corresponding row in the final entry column

Sources will be the existing PDF importers and the importers implemented in #7947.

In a second step, we want to use this functionality to clean-up bib entries. Users may select an existing bib-entry and 'enhance' it by analyzing linked pdf files. In that case, the original entry will be displayed by an additional source-column.

TODO:
- GUI glitches (text moves vertically sometimes when selecting a text)

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Review by btut — COMMENTED** [review]
I annotated some thoughts I had during implementation, any comments are highly welcome!

File: src/main/java/org/jabref/logic/importer/util/GrobidService.java [hunk 1]
```diff
@@ -59,4 +69,52 @@ public String processCitation(String rawCitation, ConsolidateCitations consolida
...
+        postFile.append(HTTP_REQUEST_BOUNDARY);
+        postFile.append("--\r\n");
+        byte[] postFileBytes = postFile.toString().getBytes();
+
+        byte[] post = new byte[preFileBytes.length + fileContent.length + postFileBytes.length];
+        System.arraycopy(preFileBytes, 0, post, 0, preFileBytes.length);
+        System.arraycopy(fileContent, 0, post, preFileBytes.length, fileContent.length);
+        System.arraycopy(postFileBytes, 0, post, preFileBytes.length + fileContent.length, postFileBytes.length);
+        return post;
+    }
```
[the remark this answers is not in the document]
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/importer/util/GrobidService.java:119]: Really ugly. Better use Jsoup: https://stackoverflow.com/a/34576290

File: src/test/java/org/jabref/logic/importer/fileformat/GrobidPdfMetadataImporterTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    @Test
+    public void testImportEntries() throws URISyntaxException {
+        Path file = Path.of(GrobidPdfMetadataImporterTest.class.getResource("LNCS-minimal.pdf").toURI());
+        List<BibEntry> bibEntries = importer.importDatabase(file, StandardCharsets.UTF_8).getDatabase().getEntries();
+
+        assertEquals(1, bibEntries.size());
+
+        BibEntry be0 = bibEntries.get(0);
+        assertEquals(Optional.of("Lastname, Firstname"), be0.getField(StandardField.AUTHOR));
+        assertEquals(Optional.of("Paper Title"), be0.getField(StandardField.TITLE));
```
[the remark this answers is not in the document]
Comment (calixtus) [reply src/test/java/org/jabref/logic/importer/fileformat/GrobidPdfMetadataImporterTest.java:52]: If I understand grobid correctly, it is basically a machine-learning search algorithm. I think the more general the search query, the more vague the result may be, as long as grobid is not yet so well trained. On the other hand, the more specific the search query, the more specific the result should be - or am I wrong?

File: src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java [hunk 3]
```diff
@@ -23,9 +23,10 @@
 
 public class GrobidCitationFetcher implements SearchBasedFetcher {
 
+    public static final String GROBID_URL = "http://grobid.jabref.org:8070";
```
[the remark this answers is not in the document]
Comment (calixtus) [reply src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java:26]: not necessarily, only if you want to make this string really customizable by the user. Working with confidential data is i think an edge case. But on the other hand, why not be prepared for edge cases, if they don't hurt to implement.

**Comment by Siedlerchr** [comment]
have you looked into the Apache Tika class? Could this be an option?
https://cwiki.apache.org/confluence/display/TIKA/GrobidJournalParser

File: src/main/java/org/jabref/logic/importer/util/GrobidService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (calixtus) [reply src/main/java/org/jabref/logic/importer/util/GrobidService.java:119]: Whole URLDownload class should probably be refactored to a fluid design too someday, but not in this PR.

File: src/test/java/org/jabref/logic/importer/fileformat/GrobidPdfMetadataImporterTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (btut) [reply src/test/java/org/jabref/logic/importer/fileformat/GrobidPdfMetadataImporterTest.java:52]: So in this case the 'search query' would be the pdf. It's quite minimalistic so I expect Grobid to detect this information. I think the tests are ok, but maybe I should add a note (as a comment) so that if it breaks in the future, people know what to do.

File: src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (btut) [reply src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java:26]: I myself would rather not send my bibliography to a third-party service (it's not even encrypted).
If we opt not to have this configurable, I already know I'll have a hosts-file entry redirecting grobid.jabref.org to my local instance ;)

**Comment by btut** [comment]
5d487d2 is a first draft of how I imagined the merge dialog.
Peek 2021-07-20 17-38

The gif shows three entries (completely different ones in this case - this is just a demonstration).
On the very left, the names of the BibTeX fields are given.
On the right, there are TextAreas that give the final result of the merge. They can be edited manually.
Users can select a whole column, which will copy all values that are available to the right (the TextAreas). Fields that are not available in that column will not be overwritten.
Clicking on a single cell copies only that cell to the right.
The clicked cells are highlighted as selected. Manually changing the entry on the right deselects all options for that field.

There is still a lot to do on the graphical end, but the functionality is mostly there and how I wanted it to be.
Open:
- clicking on a cell copies the content, but does not highlight it (only on a second click as seen in the gif)
- Merge the entry type
- Scrolling
- theming

**Comment by Siedlerchr** [comment]
Merge dialog, I like the general idea, but it has to be usable with small screens. 
Refs also #6190

File: src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (calixtus) [reply src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java:26]: Shouldn't be a problem to put this in a PreferencesTab, but I would implement this at the very end, when everything else is working, so it wont pullute the PR for now and it should be straightforward.

**Comment by calixtus** [comment]
Maybe its not even necessary to display the textfields on the right, since the information displayed is redundant to the togglebutton selected.

**Comment by btut** [comment]
> Maybe its not even necessary to display the textfields on the right, since the information displayed is redundant to the togglebutton selected.

The idea is:
- if there are errors, I want to be able to fix them immediately
- later, I would like to add the option to add fields that no importer detected (via an add button at the bottom)

**Comment by tobiasdiez** [comment]
That looks nice! Good work.

I agree with Christoph, that the merge dialog needs to work also on smaller screens. For this, the n-merge approach might be problematic and probably the user is overwhelmed anyway by the information of > 3 choices. It may be worth a consideration to merge some of the extracted metadata automatically (say based on some priority list) and then only present the user two choices. For example, merge grobid + firstpage automatically, and then have a user interface with only the options "grobid/firstpage" and "xmp". 

That brings me to the question what is the "firstpage" extractor? Is this our self-written importer that only barely works with a small selection of IEEE documents (if I remember correctly)? In this case, I would actually argue that we just remove the FirstPage extractor and replace it completely by grobid.

Based on the merging "grobid + firstpage"  I got another idea: what about automatically merging all extracted metadata (grobid + xmp) per default, with the choice in the preferences to activate the advanced merge dialog? In the end, it's a tradeoff of how quickly users can import PDFs vs the quality of the extracted metadata. And if grobid is good enough, then I would argue that it's justified to value speed over a small quality improvement. But on the other hand, there are certain user circles that have good-quality xmp metadata stored in their PDFs, so for them it would be frustrating if this is automatically overwritten by the subpar grobid metadata. For those users, the merge dialog has a lot of value and they can activate it in the preferences. What do you think?

Final small remark: For me "Grobid" is an implementation detail and most users don't know (and don't need to know) what this is. So I propose to replace it by some generic name in the UI, something like "Automatically extracted".

**Comment by btut** [comment]
> That looks nice! Good work.

Thanks :)

> I agree with Christoph, that the merge dialog needs to work also on smaller screens.

Sure. This is just a first draft, lots of details to work on. Once all that empty space is gone, the whole thing will be much smaller I hope.

> For this, the n-merge approach might be problematic and probably the user is overwhelmed anyway by the information of > 3 choices.

Actually, there will be even more than three. Grobid, Xmp, Firstpage, embedded bibtex, DOI fetcher (if any other importer detects a DOI, I would use the fetchers to get information from the DOI and add another column).

> It may be worth a consideration to merge some of the extracted metadata automatically (say based on some priority list) and then only present the user two choices. For example, merge grobid + firstpage automatically, and then have a user interface with only the options "grobid/firstpage" and "xmp".

Thats the plan for later on. I wanted to first implement the merge to get a feeling for how good Grobid works, how likely it is for XMP metadata / embedded bibtex to be present...
I though about having a user-sortable priority list in preferences, where users can also disable an importer. This could be just as overwhelming though. Do you think a static priority list defined by us would be ok?
The goal would be to step through that priority list and always keep the first value for a field we find (so if Grobid gives Author + Title and XMP gives Title and Year, we would use Author and Title from Grobid and Year from XMP, provided Gobid has higher priority).

> That brings me to the question what is the "firstpage" extractor? Is this our self-written importer that only barely works with a small selection of IEEE documents (if I remember correctly)? In this case, I would actually argue that we just remove the FirstPage extractor and replace it completely by grobid.

Yes, thats the one. AFAIK it works very well for the small set of data it works with, right? I did not check the implementation yet, will it fail if the pdf is not in IEEE format? If so, I would opt to use it if it does not fail.

> Based on the merging "grobid + firstpage" I got another idea: what about automatically merging all extracted metadata (grobid + xmp) per default, with the choice in the preferences to activate the advanced merge dialog? In the end, it's a tradeoff of how quickly users can import PDFs vs the quality of the extracted metadata. And if grobid is good enough, then I would argue that it's justified to value speed over a small quality improvement. It would be a bit of a magical feeling if you drop a PDF into JabRef, and an entry is directly created with almost-perfect metadata without any further user interaction. But on the other hand, there are certain user circles that have good-quality xmp metadata stored in their PDFs, so for them it would be frustrating if this is automatically overwritten by the subpar grobid metadata. For those users, the merge dialog has a lot of value and they can activate it in the preferences. What do you think?

See above. This sounds like an argument to make the priority list sortable.
 
> Final small remark: For me "Grobid" is an implementation detail and most users don't know (and don't need to know) what this is. So I propose to replace it by some generic name in the UI, something like "Automatically extracted".

True, but all others are 'automatically extracted' as well. Before starting this project, I didn't know about XMP either. Maybe we should just use generic names for the options (Source 1, Source 2) and show details on mouse-over?

**Comment by Siedlerchr** [comment]
Regarding the custom first page importer, it also checks if it finds a DOI on the page and then simply calls the DOI fetcher with that doi

**Comment by btut** [comment]
> Regarding the custom first page importer, it also checks if it finds a DOI on the page and then simply calls the DOI fetcher with that doi

In that case, I would disable that feature from the first-page importer and do DOI lookup last. So if any importer finds a DOI, use the DOI fetcher.

**Comment by tobiasdiez** [comment]
> Actually, there will be even more than three. Grobid, Xmp, Firstpage, embedded bibtex, DOI fetcher (if any other importer detects a DOI, I would use the fetchers to get information from the DOI and add another column).

I'm a bit worried that displaying all of this information in one screen will be overwhelming. Maybe condense it into "Extracted" (Grobid, Firstpage, DOI merged) vs "Embedded" (XMP, bibtex merged) ?

> Do you think a static priority list defined by us would be ok?

Yes! Don't make everything configurable, especially not at the beginning.

> This sounds like an argument to make the priority list sortable.

This was more an argument/idea to merge all information by default. In the end, the user doesn't care what kind of sources we used, he just wants to have a high-quality entry when he puts a pdf into JabRef.

File: src/main/java/org/jabref/logic/importer/util/GrobidService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (btut) [reply src/main/java/org/jabref/logic/importer/util/GrobidService.java]: Finally looked into Jsoup. Thanks, @Siedlerchr, for the recommendation, this is so much better.

**Comment by calixtus** [comment]
Localization test is failing

File: src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesView.java [hunk 4]
```diff
@@ -232,64 +233,67 @@ private boolean isMultilineField(Field field) {
...
-            buttonTooltip.setWrapText(true);
-            buttonTooltip.prefWidthProperty().bind(widthProperty());
-            buttonTooltip.setTextAlignment(TextAlignment.LEFT);
-            cellButton.setTooltip(buttonTooltip);
-
-            cellButton.setToggleGroup(row.toggleGroup);
-            if (row.toggleGroup.getSelectedToggle() == null) {
-                cellButton.setSelected(true);
-            }
+            DefaultTaskExecutor.runInJavaFXThread(() -> {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesView.java:236]: I would suggest adding a  short comment on why this is necessary here at this point

Comment (btut) [reply src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesView.java:236]: I'm still not entirely sure why it is necessary. Would giving the effect that we avoid be enough?

(Also, I'm sorry for not mentioning you as co-author of 4c74d51)

**Comment by calixtus** [comment]
Something idea that just came to me: In that screen recording some sources are greyed out. How about some button to collapse those empty sources to a small column of just 1em or so?

File: src/main/java/org/jabref/gui/mergeentries/DiffHighlightingEllipsingTextFlow.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+                    break;
+                case WORD:
+                    highlightedText = DiffHighlighting.generateDiffHighlighting(comparisonString.get(), fullText, " ");
+                    break;
+                case CHARACTER:
+                    highlightedText = DiffHighlighting.generateDiffHighlighting(comparisonString.get(), fullText, "");
+                    break;
+                default:
+                    throw new UnsupportedOperationException("Not implemented " + diffMode.getValue());
+            }
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/mergeentries/DiffHighlightingEllipsingTextFlow.java:142]: modern style switch expression ?
`final List<Text> hightlightedText = switch (diffMode.getValue()) ... yield List.of(text); ...` ?

**Review by calixtus — APPROVED** [review]
just one suggestion

**Comment by calixtus** [comment]
About diff: https://discourse.jabref.org/t/duplicated-entries-which-comparison-mode-do-you-use/114

**Comment by koppor** [comment]
Fixes https://github.com/koppor/jabref/issues/506

File: src/test/java/org/jabref/logic/importer/util/GrobidServiceTest.java [hunk 6]
```diff
@@ -49,10 +49,14 @@ public void processValidCitationTest() throws IOException {
         assertTrue(responseRows[1].contains("author") && responseRows[1].contains("Derwing, Tracey and Rossiter, Marian and Munro, Murray"));
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/logic/importer/util/GrobidServiceTest.java]: I don't like this assertTrue stuff here.
Isn't there any way to have an assertEquals?
 I would propose a List.of(first line text, secondlinetext) and then compare it to  the response

File: src/main/java/org/jabref/gui/externalfiles/DownloadFullTextAction.java [hunk 7]
```diff
@@ -144,8 +144,7 @@ private void addLinkedFileFromURL(BibDatabaseContext databaseContext, URL url, B
                     databaseContext,
                     Globals.TASK_EXECUTOR,
                     dialogService,
-                    preferences.getXmpPreferences(),
-                    preferences.getFilePreferences(),
+                    preferences,
```
Comment (koppor) [inline src/main/java/org/jabref/gui/externalfiles/DownloadFullTextAction.java:147]: Why was this change necessary? We should focus the preferences passed around - and not enlarge the scope.

**Review by koppor — COMMENTED** [review]
I see the introduction of `Preferences` and not using smaller-grained preferences. What was the issue? Should we introduce some more data-transfer-objects with composite preferences? I'd like to avoid to introduce a de-facto global variable `preferences` and would like to focus on small preferences classes.

File: src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesView.java [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+                .withText(MergeEntries.DiffMode::getDisplayText)
+                .install(diffMode);
+        MergeEntries.DiffMode diffModePref = preferences.getMergeDiffMode()
+                                                          .flatMap(MergeEntries.DiffMode::parse)
+                                                          .orElse(MergeEntries.DiffMode.WORD);
+        diffMode.setValue(diffModePref);
+
+        EasyBind.subscribe(this.diffMode.valueProperty(), mode -> {
+            preferences.storeMergeDiffMode(mode.name());
+        });
```
Comment (btut) [inline src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesView.java:149]: @koppor this is why I need the big preferences object.

File: src/test/java/org/jabref/logic/importer/util/GrobidServiceTest.java (hunk 6 above)
Comment (btut) [reply src/test/java/org/jabref/logic/importer/util/GrobidServiceTest.java]: True, but not exactly related to this PR, so I opened #8003.

File: src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesView.java (hunk 8 above)
Comment (calixtus) [reply src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesView.java:149]: @DominikVoigt and me are looking into reworking preferences handling. Should be merged anyways to move forward and leave the preferences problem to a more general solution.

**Review by calixtus — APPROVED** [review]
Some remarks were made in PR #8002 .
After fixing them, this should be ready too.
Otherweise lgtm.

**Comment by calixtus** [comment]
Before merging this PR, merge #8002 !
This one closes #8002 .

File: CHANGELOG.md [hunk 9]
```diff
@@ -17,6 +17,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We added unprotect_terms to the list of bracketed pattern modifiers [#7826](https://github.com/JabRef/jabref/pull/7960)
 - We added a dialog that allows to parse metadata from linked pdfs. [#7929](https://github.com/JabRef/jabref/pull/7929)
 - We added an icon picker in group edit dialog. [#6142](https://github.com/JabRef/jabref/issues/6142)
+- We added a preference to Opt-In to Grobid usage. [8002](https://github.com/JabRef/jabref/pull/8002)
```
Comment (Siedlerchr) [inline CHANGELOG.md:20]: JabRef's online metadata extraction (Grobid) service usage

Comment (btut) [reply CHANGELOG.md]: Fixed in #8002 and merged into this PR.

File: src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesViewModel.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+    public ObjectProperty<BibEntry> mergedEntryProperty() {
+        return mergedEntry;
+    }
+
+    public ListProperty<String> failedSuppliersProperty() {
+        return failedSuppliers;
+    }
+
+    public static class Entry {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesViewModel.java:78]: Maybe rename to MergeEntry, because otherwise it's confusing with BibEntry and Entry

Comment (btut) [reply src/main/java/org/jabref/gui/mergeentries/MultiMergeEntriesViewModel.java]: Makes sense, but I chose EntrySource instead because it is one option for the merge, not the merged entry.

**Review by Siedlerchr — APPROVED** [review]

**Comment by calixtus** [comment]
Two green checkmarks. Codacy is complaining about 4-space tabs instead of 2, but in every other file its done like here. So merging now. 🎉
